internal/security/bootstrapper/command/setupacl/acltokens.go

```diff
@@ -0,0 +1,413 @@
+/*******************************************************************************
+ * Copyright 2021 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ *******************************************************************************/
+
+package setupacl
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-bootstrap/v2/bootstrap/startup"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients"
+)
+
+// AgentTokenType is the type of token to be set on the Consul agent
+type AgentTokenType string
+
+const (
+	/*
+	 * The following are available agent token types that can be used for setting the token to Consul's agent
+	 * For the details for each type, see reference https://www.consul.io/commands/acl/set-agent-token#token-types
+	 */
+	// DefaultType is agent token type "default" to be set
+	DefaultType AgentTokenType = "default"
+	// AgentType is agent token type "agent" to be set
+	AgentType AgentTokenType = "agent"
+	// MasterType is agent token type "master" to be set
+	MasterType AgentTokenType = "master"
+	// ReplicationType is agent token type "replication" to be set
+	ReplicationType AgentTokenType = "replication"
+
+	// consul API related:
+	consulTokenHeader      = "X-Consul-Token"
+	consulCheckAgentAPI    = "/v1/agent/self"
+	consulSetAgentTokenAPI = "/v1/agent/token/%s"
+	consulListTokensAPI    = "/v1/acl/tokens"
+	consulCreateTokenAPI   = "/v1/acl/token"
+	// RUD: Read Update Delete
+	consulTokenRUDAPI = "/v1/acl/token/%s"
+)
+
+// isACLTokenPersistent checks Consul agent's configuration property for EnablePersistence of ACLTokens
+// it returns true if the token persistence is enabled; false otherwise
+// once ACL rules are enforced, this call requires at least agent read permission and hence we use
+// the bootstrap ACL token as that's the only token available before creating an agent token
+// this determines whether we need to re-set the agent token every time Consul agent is restarted
```

Comment:
Can't we just force EnablePersistence = true by setting in the config file?  Or if we need to check it, just outright fail if the precondition is not met?

Reply:
we could, but i think this way is more robust. as I learned you can dynamically change conusl configure on the fly and reload.

Reply:
https://www.consul.io/commands/reload